nilearn/image/resampling.py

```diff
@@ -657,42 +544,235 @@ def resample_img(
+
+    Parameters
+    ----------
+    target_affine : numpy.ndarray
+        If specified, the image is resampled corresponding to this new affine.
+        target_affine can be a 3x3 or a 4x4 matrix.
+
+    target_shape : tuple or list
+        If specified, the image will be resized to match this new shape.
+        len(target_shape) must be equal to 3.
+        If target_shape is specified, a target_affine of shape (4, 4)
+        must also be given.
+
+    affine : numpy.ndarray
+        Affine of the input image.
+
+    data : numpy.ndarray
+        data of the input image.
+    """
+    if target_affine.shape == (3, 3):
+        # Embed target_affine in 4x4 shape
+        missing_offset = True
+        target_affine_tmp = np.eye(4)
+        target_affine_tmp[:3, :3] = target_affine
+        target_affine = target_affine_tmp
+    else:
+        missing_offset = False
+        target_affine = target_affine.copy()
+
+    transform_affine = np.linalg.inv(target_affine).dot(affine)
+    (xmin, xmax), (ymin, ymax), (zmin, zmax) = get_bounds(
+        data.shape[:3], transform_affine
+    )
+
+    # if target_affine is (3, 3),
+    # then calculate offset from bounding box and update bounding box
+    # to be in the voxel coordinates of the calculated 4x4 affine
+    if missing_offset:
+        offset = target_affine[:3, :3].dot([xmin, ymin, zmin])
+        target_affine[:3, 3] = offset
+        (xmin, xmax), (ymin, ymax), (zmin, zmax) = (
+            (0, xmax - xmin),
+            (0, ymax - ymin),
+            (0, zmax - zmin),
+        )
+
+    # Check whether transformed data is actually within the FOV
+    # of the target affine
+    if xmax < 0 or ymax < 0 or zmax < 0:
+        raise BoundingBoxError(
+            "The field of view given by the target affine does not "
+            "contain any of the data."
+        )
+
+    # if target_shape is not given
+    # (always the case with 3x3 transformation matrix
+    # and sometimes the case with 4x4 transformation matrix),
+    # then set it to contain the bounding box by a margin of 1 voxel
+    if target_shape is None:
+        target_shape = (
+            int(np.ceil(xmax)) + 1,
+            int(np.ceil(ymax)) + 1,
+            int(np.ceil(zmax)) + 1,
+        )
+
+    return target_affine, target_shape
+
+
+def _clip_img(data, resampled_data, clip):
+    """Force resampled data to have a range contained in the original data \
+    preventing ringing artefact.
+
+    We need to add zero as a value considered for clipping,
+    as it appears in padding images.
+    """
+    if not clip:
+        return resampled_data
+    vmin = min(data.min(), 0)
+    vmax = max(data.max(), 0)
+    resampled_data.clip(vmin, vmax, out=resampled_data)
+    return resampled_data
+
+
+def _interpolation_order(interpolation):
+    if interpolation == "continuous":
+        return 3
+    elif interpolation == "linear":
+        return 1
+    elif interpolation == "nearest":
+        return 0
```

Comment:
maybe `{"continuous": 3, "linear": 1, "nearest": 0}[interpolation]`? to make it clear there is no default case and we don't want to return None